src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -970,13 +1375,196 @@ fn slice_pat_covered_by_constructor<'tcx>(
+                let add_endpoint = |points: &mut FxHashMap<_, _>, x, endpoint| {
+                    points.entry(x).and_modify(|ex_x| {
+                        if *ex_x != endpoint {
+                            *ex_x = Endpoint::Both
+                        }
+                    }).or_insert(endpoint);
+                };
+                let add_endpoints = |points: &mut FxHashMap<_, _>, lo, hi| {
+                    // Insert the endpoints, taking care to keep track of to
+                    // which endpoints a point corresponds.
+                    add_endpoint(points, lo, Endpoint::Start);
+                    add_endpoint(points, hi, Endpoint::End);
+                };
+                let (lo, hi) = (*ctor_range.range.start(), *ctor_range.range.end());
+                add_endpoints(&mut points, lo, hi);
+                // We're going to iterate through every row pattern, adding endpoints in.
+                for row in m.iter() {
+                    if let Some(r) = IntRange::from_pat(tcx, row[0]) {
+                        // We're only interested in endpoints that lie (at least partially)
+                        // within the subrange domain.
+                        if let Some(r) = ctor_range.intersection(&r) {
+                            let (r_lo, r_hi) = r.range.into_inner();
+                            add_endpoints(&mut points, r_lo, r_hi);
+                        }
+                    }
+                }
+
+                // The patterns were iterated in an arbitrary order (i.e. in the order the user
+                // wrote them), so we need to make sure our endpoints are sorted.
+                let mut points: Vec<(u128, Endpoint)> = points.into_iter().collect();
+                points.sort_unstable_by_key(|(x, _)| *x);
+                let mut points = points.into_iter();
+                let mut a = points.next().unwrap();
+
+                // Iterate through pairs of points, adding the subranges to `split_ctors`.
+                // We have to be careful about the orientation of the points as endpoints, to make
+                // sure we're enumerating precisely the correct ranges. Too few and the matching is
+                // actually incorrect. Too many and our diagnostics are poorer. This involves some
+                // case analysis.
+                // In essence, we need to ensure that every time the set of row-ranges that are
+                // overlapping changes (as we go through the values covered by the ranges), we split
+                // into a new subrange.
+                while let Some(b) = points.next() {
+                    // a < b (strictly)
+                    if let Endpoint::Both = a.1 {
+                        split_ctors.push(IntRange::range_to_ctor(tcx, ty, a.0..=a.0));
+                    }
+                    let c = match a.1 {
+                        Endpoint::Start => a.0,
+                        Endpoint::End | Endpoint::Both => a.0 + 1,
```

Comment:
How does this handle integer overflow (i.e. `a.0 = uint128::MAX`)? worth a test case, at least.

Reply:
Oh, that's a good point 👍

Reply:
Other than that and the duplication, this looks good.